components/eamxx/scripts/atm_manip.py

```diff
@@ -250,71 +136,100 @@ def atm_config_chg_impl(xml_parent, xml_node, change, all_matches=False):
+    >>> atm_config_chg_impl(tree,'prop1=three', all_matches=True)
+    True
+    >>> [item.text for item in get_xml_nodes(tree,'prop1')]
+    ['three', 'three']
     >>> ################ TEST APPEND += #################
-    >>> atm_config_chg_impl(None,tree,'a+=4')
-    (True, True)
-    >>> get_xml_node(None,tree,'a')[0].text
+    >>> atm_config_chg_impl(tree,'a+=4')
+    True
+    >>> get_xml_nodes(tree,'a')[0].text
     '1,2,3, 4'
     >>> ################ ERROR, append to non-array and non-string
-    >>> atm_config_chg_impl(None,tree,'c+=2')
+    >>> atm_config_chg_impl(tree,'c+=2')
     Traceback (most recent call last):
     SystemExit: ERROR: Error! Can only append with array and string types.
         - name: c
         - type: int
     >>> ################ Append to string ##################
-    >>> atm_config_chg_impl(None,tree,'d+=two')
-    (True, True)
-    >>> get_xml_node(None,tree,'d')[0].text
+    >>> atm_config_chg_impl(tree,'d+=two')
+    True
+    >>> get_xml_nodes(tree,'d')[0].text
     'onetwo'
     >>> ################ Append to array(string) ##################
-    >>> atm_config_chg_impl(None,tree,'e+=two')
-    (True, True)
-    >>> get_xml_node(None,tree,'e')[0].text
+    >>> atm_config_chg_impl(tree,'e+=two')
+    True
+    >>> get_xml_nodes(tree,'e')[0].text
     'one, two'
     """
+    node_name, new_value, append_this = parse_change(change)
+    matches = get_xml_nodes(xml_root, node_name)
+
+    expect(len(matches) > 0, f"{node_name} did not match any items")
+    if len(matches) > 1 and not all_matches:
+        parent_map = create_parent_map(xml_root)
+        error_str = ""
+        for node in matches:
+            parents = get_parents(node, parent_map)
+            name = "::".join(e.tag for e in parents) + "::" + node.tag
+            error_str += "  " + name + "\n"
+
+        expect(False, f"{node_name} is ambiguous, matches:\n{error_str}")
 
     any_change = False
-    node_found = False
-    if all_matches and len(xml_node)>0:
-        # We have to go through the whole tree, since we need to apply
-        # the changes to all nodes matching the node name
-        for elem in xml_node:
-            found_here, changed_here = atm_config_chg_impl(xml_node,elem,change, all_matches)
-            any_change |= changed_here
-            node_found |= found_here
-    else:
-        node_name,new_value,append_this = parse_change(change)
-        node, __ = get_xml_node(xml_parent,node_name,allow_not_found=all_matches)
+    for node in matches:
+        any_change |= apply_change(node, new_value, append_this)
+
+    return any_change
 
-        node_found = node is not None
-        if node is not None:
-            any_change = apply_change (node,new_value,append_this)
+###############################################################################
+def create_parent_map(root):
+###############################################################################
+    return {c: p for p in root.iter() for c in p}
```

Comment:
The Python is strong in you. The double list comprehension is giving me headaches.